### PR TITLE
Slightly optimize Array/Tuple/Map

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

It is less important for Tuple/Map since they are unlikely to be as large as arrays, but still.

Follow-up for: #34074

### CODE REVIEW

File: src/DataTypes/Serializations/SerializationTuple.cpp [hunk 1]
```diff
@@ -44,9 +44,11 @@ void SerializationTuple::deserializeBinary(Field & field, ReadBuffer & istr) con
...
 
-    Tuple tuple(size);
+    field = Tuple();
+    Tuple & tuple = get<Tuple &>(field);
+    tuple.reserve(size);
     for (const auto i : collections::range(0, size))
-        elems[i]->deserializeBinary(tuple[i], istr);
+        elems[i]->deserializeBinary(tuple.emplace_back(), istr);
 
     field = tuple;
```
Comment (kitaisreal) [inline src/DataTypes/Serializations/SerializationTuple.cpp:53]: Consider to use `std::move` or remove this line.

File: src/Columns/ColumnTuple.cpp [hunk 2]
```diff
@@ -101,17 +101,21 @@ MutableColumnPtr ColumnTuple::cloneResized(size_t new_size) const
...
-    Tuple tuple(tuple_size);
-    for (const auto i : collections::range(0, tuple_size))
-        columns[i]->get(n, tuple[i]);
 
-    res = tuple;
+    res = Tuple();
+    Tuple & res_tuple = DB::get<Tuple &>(res);
+    res_tuple.reserve(tuple_size);
+
+    for (const auto i : collections::range(0, tuple_size))
```
Comment (kitaisreal) [inline src/Columns/ColumnTuple.cpp:117]: `Consider to use simple loop without collections::range.`

**Review by kitaisreal — APPROVED** [review]

File: src/Columns/ColumnTuple.cpp (hunk 2 above)
Comment (azat) [reply src/Columns/ColumnTuple.cpp:117]: I've leaved it as-is in the surrounding code, but I'm ok to change it.

File: src/DataTypes/Serializations/SerializationTuple.cpp (hunk 1 above)
Comment (azat) [reply src/DataTypes/Serializations/SerializationTuple.cpp:53]: Good catch.

**Comment by kitaisreal** [comment]
@Mergifyio update

**Pull request merged**